private_transactions/src/encryptor.rs

```diff
@@ -0,0 +1,318 @@
+		let passwords = self.config.passwords.clone();
+		for password in passwords {
+			if let Ok(()) = accounts.unlock_account_temporarily(account.clone(), password) {
+				return Ok(true);
+			}
+		}
+		Ok(false)
+	}
+
+	fn sign_contract_address(&self, contract_address: &Address, accounts: Arc<AccountProvider>) -> Result<Signature, PrivateTransactionError> {
+		// key id in SS is H256 && we have H160 here => expand with assitional zeros
+		let contract_address_extended: H256 = contract_address.into();
+		let key_server_account = self.config.key_server_account.ok_or_else(|| PrivateTransactionError::KeyServerAccountNotSet)?;
+		if let Ok(true) = self.unlock_account(&key_server_account, accounts.clone()) {
+			Ok(accounts.sign(key_server_account.clone(), None, H256::from_slice(&contract_address_extended))?)
+		} else {
+			trace!("Cannot unlock account");
+			Err(PrivateTransactionError::Encrypt("Cannot unlock account".into()))
+		}
+	}
+}
+
+impl Encryptor for SecretStoreEncryptor {
+	fn encrypt(
+		&self,
+		contract_address: &Address,
+		accounts: Arc<AccountProvider>,
+		initialisation_vector: &H128,
+		plain_data: &[u8]
+	) -> Result<Bytes, PrivateTransactionError> {
+		// retrieve the key, try to generate it if it doesn't exist yet
+		let key = match self.retrieve_key("", false, contract_address, accounts.clone()) {
+			Ok(key) => Ok(key),
+			Err(PrivateTransactionError::EncryptionKeyNotFound(_)) => {
+				trace!("Key for account wasnt found in sstore. Creating. Address: {:?}", contract_address);
+				self.retrieve_key(&format!("/{}", self.config.threshold), true, contract_address, accounts.clone())
+			}
+			Err(err) => Err(err),
+		}?;
+
+		// encrypt data
+		let mut cypher = Vec::with_capacity(plain_data.len() + initialisation_vector.len());
+		cypher.extend(repeat(0).take(plain_data.len()));
+		ethcrypto::aes::encrypt(&key, initialisation_vector, plain_data, &mut cypher);
+		cypher.extend_from_slice(&initialisation_vector);
+
+		Ok(cypher)
+	}
+
+	/// Decrypt data using previously generated contract key.
+	fn decrypt(
+		&self,
+		contract_address: &Address,
+		accounts: Arc<AccountProvider>,
+		cypher: &[u8]
+	) -> Result<Bytes, PrivateTransactionError> {
+		// initialization vector takes INIT_VEC_LEN bytes
+		let cypher_len = cypher.len();
+		if cypher_len < INIT_VEC_LEN {
+			return Err(PrivateTransactionError::Decrypt("Invalid cypher".into()));
+		}
+
+		// retrieve existing key
+		let key = self.retrieve_key("", false, contract_address, accounts)?;
+
+		// use symmetric decryption to decrypt document
+		let (cypher, iv) = cypher.split_at(cypher_len - INIT_VEC_LEN);
+		let mut plain_data = Vec::with_capacity(cypher_len - INIT_VEC_LEN);
+		plain_data.extend(repeat(0).take(cypher_len - INIT_VEC_LEN));
+		ethcrypto::aes::decrypt(&key, &iv, cypher, &mut plain_data);
+
+		Ok(plain_data)
+	}
+}
+
+/// Dummy encryptor.
+#[derive(Default)]
+pub struct DummyEncryptor;
```

Comment:
this structure is `pub` and therefore a part of the crate. It's used in tests located in multiple files. Maybe instead of exposing it a part of crate, move all tests to `tests/` directory and define this structure over there

Reply:
I've moved tests to the separate directory, it's good for code's structure anyway. But DummyEncryptor left public for the crate, because it's used outside (in sync tests)